### PR TITLE
UCM/Allocator: Add new version of the memory allocator.

### PR DESCRIPTION
**Pull request by shssf**
Related issue #1047 
Added configure script parameter:
```
  --with-allocator=NAME   Build UCX with predefined memory allocator. The     
                          supported values are: ptmalloc283, ptmalloc286.     
                          Default: ptmalloc286                                
```

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2816/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/248/ for details.

File: src/ucm/malloc/malloc_hook.c [hunk 1]
```diff
@@ -17,7 +17,13 @@
 #include <ucm/api/ucm.h>
 #include <ucm/event/event.h>
 #include <ucm/mmap/mmap.h>
-#include <ucm/ptmalloc3/malloc-2.8.3.h>
+#if defined(HAVE_UCM_PTMALLOC283)
```
Comment (yosefe) [inline src/ucm/malloc/malloc_hook.c]: why not #if HAVE_UCM_PTMALLOC283 / #ifdef HAVE_UCM_PTMALLOC283 ?
also - we might want to put these lines in separate file e.g ucm/malloc/allocator.h

File: src/ucm/ptmalloc286/malloc-2.8.6.h [hunk 2]
```diff
@@ -0,0 +1,622 @@
+/*
+  Default header file for malloc-2.8.x, written by Doug Lea
+  and released to the public domain, as explained at
+  http://creativecommons.org/publicdomain/zero/1.0/ 
+
```
Comment (shamisp) [inline src/ucm/ptmalloc286/malloc-2.8.6.h]: did you check the licensing for it ?

**Comment by shamisp** [comment]
what is the difference between 283 and 286 ?

**Comment by yosefe** [comment]
better performance when allocating from multiple threads

**Comment by shamisp** [comment]
did they have some sort of bugfix ? Just curious ...

**Comment by shssf** [comment]
@shamisp I'm not sure there is a bug in 2.8.3. They just improve locks performance in modern version.
Looking in the history in malloc.c (at the end of the file)
<pre>
History:
    v2.8.6 Wed Aug 29 06:57:58 2012  Doug Lea
      * fix bad comparison in dlposix_memalign
      * don't reuse adjusted asize in sys_alloc
      * add LOCK_AT_FORK -- thanks to Kirill Artamonov for the suggestion
      * reduce compiler warnings -- thanks to all who reported/suggested these

    v2.8.5 Sun May 22 10:26:02 2011  Doug Lea  (dl at gee)
      * Always perform unlink checks unless INSECURE
      * Add posix_memalign.
      * Improve realloc to expand in more cases; expose realloc_in_place.
        Thanks to Peter Buhr for the suggestion.
      * Add footprint_limit, inspect_all, bulk_free. Thanks
        to Barry Hayes and others for the suggestions.
<b>      * Internal refactorings to avoid calls while holding locks
      * Use non-reentrant locks by default. Thanks to Roland McGrath
        for the suggestion.</b>
      * Small fixes to mspace_destroy, reset_on_error.
      * Various configuration extensions/changes. Thanks
         to all who contributed these.

    V2.8.4a Thu Apr 28 14:39:43 2011 (dl at gee.cs.oswego.edu)
      * Update Creative Commons URL

    V2.8.4 Wed May 27 09:56:23 2009  Doug Lea  (dl at gee)
      * Use zeros instead of prev foot for is_mmapped
      * Add mspace_track_large_chunks; thanks to Jean Brouwers
      * Fix set_inuse in internal_realloc; thanks to Jean Brouwers
      * Fix insufficient sys_alloc padding when using 16byte alignment
      * Fix bad error check in mspace_footprint
      * Adaptations for ptmalloc; thanks to Wolfram Gloger.
      * Reentrant spin locks; thanks to Earl Chew and others
      * Win32 improvements; thanks to Niall Douglas and Earl Chew
      * Add NO_SEGMENT_TRAVERSAL and MAX_RELEASE_CHECK_RATE options
      * Extension hook in malloc_state
      * Various small adjustments to reduce warnings on some compilers
      * Various configuration extensions/changes for more platforms. Thanks
         to all who contributed these.

    V2.8.3 Thu Sep 22 11:16:32 2005  Doug Lea  (dl at gee)
      * Add max_footprint functions
      * Ensure all appropriate literals are size_t
      * Fix conditional compilation problem for some #define settings
etc...
</pre>

File: src/ucm/ptmalloc286/malloc-2.8.6.h (hunk 2 above)
Comment (shssf) [inline src/ucm/ptmalloc286/malloc-2.8.6.h]: It is the same as for version 2.8.3

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2210/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2823/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/256/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2216/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp pls review as well

**Comment by shamisp** [comment]
👍

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/260/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/262/ for details.

**Pull request merged**